tawazi/dag.py

```diff
@@ -414,23 +487,224 @@ def get_highest_priority_nodes(nodes: List[ExecNode]) -> List[ExecNode]:
+    #           profile: the time it took for every single
+    #           exec_nodes: detailed object containing all exec_nodes hence their results (it should be used only for advanced usages)
+    #           args passed to the pipeline
+    # def execute(self, *args, **kwargs, twz_nodes=None, profile=False):
+
+    def _make_call_xn_dict(self, *args: Any) -> Dict[IdentityHash, ExecNode]:
+        """
+        Generate the calling ExecNode dict.
+        This is an ExecNode dict that will contain the ExecNodes that will be executed (hence modified) by the DAG scheduler.
+        This takes into consideration:
+         1. deep copying the ExecNodes
+         2. filling the arguments of the call
+         3. skipping the copy for setup ExecNodes
+        """
+        # NOTE: there is a double deep copy, this is the 1st,
+        #  the 2nd happens in DAG.execute(...) which will be removed in Tawazi 0.4
+        # 1. deepcopy the node_dict because it will be modified by the DAG's execution
+        call_xn_dict = DAG._deepcopy_non_setup_x_nodes(self.node_dict)
+
+        # 2. parse the input arguments of the pipeline
+        # 2.1 default valued arguments can be skipped and not provided!
+        # note: if not enough arguments are provided then the code will fail inside the DAG's execution through the raise_err lambda
+        if args:
+            # 2.2 can't provide more than enough arguments
+            if len(args) > len(self.input_ids):
+                raise TypeError(
+                    f"The DAG takes a maximum of {len(self.input_ids)} arguments. {len(args)} arguments provided"
+                )
+
+            # 2.3 modify ExecNodes corresponding to input ArgExecNodes
+            for ind_arg, arg in enumerate(args):
+                node_id = self.input_ids[ind_arg]
+
+                call_xn_dict[node_id].result = arg
+
+        return call_xn_dict
+
+    def _get_return_values(
+        self, xn_dict: Dict[IdentityHash, ExecNode]
+    ) -> Union[Any, Tuple[Any], List[Any]]:
+        """
+        Extract the return value/values from the output of the DAG's scheduler!
+
+        Args:
+            xn_dict (Dict[IdentityHash, ExecNode]): _description_
+
+        Raises:
+            TawaziTypeError: _description_
+
+        Returns:
+            Union[Any, Tuple[Any], List[Any]]: _description_
+        """
+        if self.return_ids is None:
+            return None
+        if isinstance(self.return_ids, IdentityHash):
+            return filter_NoVal(xn_dict[self.return_ids].result)
+        gen = (filter_NoVal(xn_dict[ren_id].result) for ren_id in self.return_ids)
+        if isinstance(self.return_ids, tuple):
+            return tuple(gen)
+        if isinstance(self.return_ids, list):
+            return list(gen)
+        raise TawaziTypeError("Return type for the DAG can only be a single value, Tuple or List")
+
+    # NOTE: this function should be used in case there was a bizarre behavior noticed during the
```

Comment:
Of course we can add an example in the README just for it... This is a good note! 
what can we name it ? The most importantce of this method is that it doesn't use the scheduler... so no parallel processing will ever happen. This should only be used as a fallback in case the normal __call__ is failing for some reason. 

We can even make it private at the moment ! we can leave it for us only as an experimental interface to try when debugging tawazi in case there is a bug that we discover in the scheduler

Reply:
what do you think ?

Reply:
I think leaving it private is clearly not a bad idea, at least until we determine its real purpose